File: app/src/main/java/com/compuspulse/viewmodel/SearchViewModel.java
package com.compuspulse.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.compuspulse.data.remote.model.Article;
import com.compuspulse.data.repository.RepositoryCallback;
import com.compuspulse.data.repository.SearchRepository;

import java.util.ArrayList;
import java.util.List;

public class SearchViewModel extends ViewModel {

    private final SearchRepository repository = new SearchRepository();
    private final MutableLiveData<List<Article>> resultList = new MutableLiveData<>(new ArrayList<>());
    private final MutableLiveData<Boolean> loading = new MutableLiveData<>(false);
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();

    public LiveData<List<Article>> getResultList() {
        return resultList;
    }

    public LiveData<Boolean> getLoading() {
        return loading;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public void search(String keyword) {
        loading.setValue(true);
        repository.searchArticles(keyword, new RepositoryCallback<List<Article>>() {
            @Override
            public void onSuccess(List<Article> data) {
                resultList.setValue(data == null ? new ArrayList<>() : data);
                loading.setValue(false);
            }

            @Override
            public void onError(String message) {
                resultList.setValue(new ArrayList<>());
                errorMessage.setValue(message);
                loading.setValue(false);
            }
        });
    }
}

File: app/src/main/java/com/compuspulse/viewmodel/SearchViewModel.kt
package com.compuspulse.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.compuspulse.data.remote.model.Article
import com.compuspulse.data.repository.SearchRepository
import kotlinx.coroutines.launch

class SearchViewModel : ViewModel() {

    private val repository = SearchRepository()
    private val resultList = MutableLiveData<List<Article>>(ArrayList())
    private val loading = MutableLiveData(false)
    private val errorMessage = MutableLiveData<String>()

    fun getResultList(): LiveData<List<Article>> = resultList

    fun getLoading(): LiveData<Boolean> = loading

    fun getErrorMessage(): LiveData<String> = errorMessage

    fun search(keyword: String) {
        loading.value = true
        viewModelScope.launch {
            repository.searchArticles(keyword)
                .onSuccess { data ->
                    resultList.value = data
                    loading.value = false
                }
                .onFailure { throwable ->
                    resultList.value = ArrayList()
                    errorMessage.value = throwable.message ?: "搜索失败"
                    loading.value = false
                }
        }
    }
}
